File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64
from std_srvs.srv import SetBool


# Inherit from Node obj
class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.counter = Int64()
        self.counter.data = 0
        # create_publisher(data_type_of_topic, topic_name, queue_size)
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.subscriber_ = self.create_subscription(Int64, "number", self.count_number, 10)
        # self.timer_ = self.create_timer(0.5, self.publish_number)
        self.server_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter)
        # For debugging
        self.get_logger().info("Number Counter has been started")
    
    def count_number(self, num):
        self.counter.data += num.data
        # self.get_logger().info("Number Counter: "+str(self.counter.data))
        self.publisher_.publish(self.counter)
        self.get_logger().info("Counter: "+str(self.counter.data))

    def callback_reset_counter(self, request, response):
        if (request.data):
            self.counter.data = 0
            response.success = True
            response.message = "Counter has been initialized."
        else:
            response.success = False
            response.message = "Counter has NOT been initialized."
        return response

def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__=="__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"
#include "std_srvs/srv/set_bool.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), counter_(0)
    {
        //counter_.data = 0;
        subscriber_ = this->create_subscription<std_msgs::msg::Int64>(
            "number", 10,
            std::bind(&NumberCounterNode::callbackCountNumber, this, std::placeholders::_1));
        publisher_ = this->create_publisher<std_msgs::msg::Int64>("number_count", 10);
        server_ = this->create_service<std_srvs::srv::SetBool>(
            "reset_counter", std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
            
        RCLCPP_INFO(this->get_logger(), "Number Counter has been started.");
    }

private:
    
    void callbackCountNumber(const std_msgs::msg::Int64::SharedPtr num)
    {
        counter_ += num->data;
        // -> is equal to (*...). . Here I have the pointer to num, so I have to use ->
        //RCLCPP_INFO(this->get_logger(), std::to_string(counter_.data));
        auto new_msg = std_msgs::msg::Int64();
        new_msg.data = counter_;
        publisher_->publish(new_msg);
        RCLCPP_INFO(this->get_logger(), "Counter: %d", new_msg.data);
    }
    
    void callbackResetCounter(const std_srvs::srv::SetBool_Request::SharedPtr request,
                              const std_srvs::srv::SetBool_Response::SharedPtr response)
    {   
        
        if (request->data)
        {
            counter_ = 0;
            response->success = true;
        }
        else{
            response->success = false;
        }
        RCLCPP_INFO(this->get_logger(), "Reset Counter: %d", request->data);
        
    }
    

    // Declare
    //std_msgs::msg::Int64 counter_;
    int counter_;
    rclcpp::Subscription<std_msgs::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<std_srvs::srv::SetBool>::SharedPtr server_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}